versions/3.1.1.md

```diff
@@ -2722,8 +2722,7 @@ MyResponseType:
   - $ref: '#/components/schemas/Lizard'
 ```
 
-which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`.  In this case, a discriminator MAY act as a "hint" to shortcut validation and selection of the matching schema which may be a costly operation, depending on the complexity of the schema. We can then describe exactly which field tells us which schema to use:
-
+which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Evaluating a `oneOf` can be a costly operation, so `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the evaluation more efficient and provide better error messaging. We can specify the exact field that tells us which schema is expected to match the instance:
```

Comment:
I'm a little hung up on this paragraph's use of "evaluation" -- does that mean "validation" and if so will that be clear to (knowledgable) readers? Also, it's not clear to mean that the discriminator makes _validation_ of a `oneOf` schema more efficient -- you still have to check _all_ of the schemas to know that exactly one matches. The value of the discriminator is that it can make serialization/deserialization more efficient.

Here's a revision that I think is clearer:
```suggestion
which means the payload _MUST_, by validation, match exactly one of the schemas described by `Cat`, `Dog`, or `Lizard`. Deserialization of a `oneOf` can be a costly operation, as it requires determining which schema matches the payload and thus should be used in deserialization. This problem also exists for `anyOf` schemas. A `discriminator` MAY be used as a "hint" to improve the efficiency of selection of the matching schema. The `discriminator` keyword cannot change the validation result of the `oneOf`, it can only help make the deserialization more efficient and provide better error messaging. We can specify the exact field that tells us which schema is expected to match the instance:
```

Reply:
> I'm a little hung up on this paragraph's use of "evaluation" -- does that mean "validation"

Generally, in JSON Schema land, we use the term "evaluation" rather than "validation" to acknowledge that there are multiple ways to evaluate a schema and not all of them are validation. Other types of evaluations include code generation, documentation generation, form generation, etc. However, I think in this case we're really just talking about validation, so it might be better to just say "validation".

> it's not clear to mean that the discriminator makes validation of a `oneOf` schema more efficient -- you still have to check all of the schemas to know that exactly one matches.

In most cases, you need to validate every option. The one case where you can short circuit is when you find two matching schemas. Once you find the second passing schema, you know validation will fail and you don't need to check the rest. `disciminator` can help you get there more quickly by checking the one you expect to fail first.

* Pass
* Fail
* Fail
* Fail
* Pass <- discriminator

Assuming the schemas will be evaluated in order, without a discriminator, you would have to check every schema. With a disciminator, you could check that one first and then when evaluating the rest in order, you would find the extra match right away and could skip evaluating the rest of the schemas.

It requires a bit of luck and a very specific scenario, but it's possible. The only reason I included this line was to keep the spirit of the original text, but I'm happy to take it out. It's technically correct, but the possibility of optimization is so small and so rare, it's not really worth mentioning. `anyOf` could provide a level of optimization worth mentioning, but no one uses it, which is why I used `oneOf` instead.

Reply:
> With a disciminator, you could check that one first and then when evaluating the rest in order, you would find the extra match right away and could skip evaluating the rest of the schemas.

Ah, great point. I did not think of this, but now I see how a discriminator can also make validation more efficient.